ssh-import-id.1
.TH ssh\-import\-id 1 "4 Feb 2013" ssh\-import "ssh\-import"
.SH NAME
\fBssh\-import\-id\fP \- retrieve one or more public keys from a public keyserver and append them to the current user's authorized_keys file (or some other specified file)

.SH SYNOPSIS
.BI "ssh\-import\-id [options] USER_ID_1 [gh:USER_ID_2] ... [lp:USER_ID_n]

.BI "ssh\-import\-id\-lp USER_ID_1 [USER_ID_2] ... [USER_ID_n]

.BI "ssh\-import\-id\-gh USER_ID_1 [USER_ID_2] ... [USER_ID_n]

.SH OPTIONS
 \-h | \-\-help		usage
 \-o | \-\-output F	write output to file 'F' (default ~/.ssh/authorized_keys, use "\-" for standard out)
 \-r | \-\-remove  	remove keys from authorized keys file 'F'
 PROTO:USER_ID		Protocol can be 'lp' for Launchpad.net, or 'gh' for Github.com

.SH DESCRIPTION
This utility will securely contact a public keyserver and retrieve one or more user's public keys, and append these to the current user's \fI~/.ssh/authorized_keys\fP file, standard output or any other specified output file.

User IDs can be prepended by a protocol:

 - \fBlp:\fP to use \fIhttps://launchpad.net/~%s/+sshkeys\fP
 - \fBgh:\fP to use \fIhttps://api.github.com/users/%s/keys\fP

If the protocol is not explicitly specified, then \fBssh-import-id\fP will read a URL variable string from \fI/etc/ssh/ssh_import_id\fP as installed by your package manager and configured by your system administrator.  You can override this locally by exporting the string you want in a URL environment variable.  If all of these are empty, then the protocol is assumed to be "lp:", which was the original target implementation of this tool.

Any keys added will be "labled" with a trailing comment, "# ssh-import-id PROTO:USER_ID".  Revoking keys will first look for lines in the authorized keys file ending with that label.

.SH SEE ALSO
\fIssh\fP(1)

.SH FILES
\fI/etc/ssh/ssh_import_id\fP

.SH AUTHOR
This manpage and the utility was written by Dustin Kirkland <dustin.kirkland@gmail.com> for Ubuntu systems (but may be used by others).  Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 3 published by the Free Software Foundation.

On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL.


debian/changelog
ssh-import-id (3.13-0ubuntu1) raring; urgency=low

  * debian/rules:
    - fix FTBFS, no need to syntax check setup.py

 -- Dustin Kirkland <kirkland@ubuntu.com>  Fri, 15 Feb 2013 16:58:04 -0600

ssh-import-id (3.12-0ubuntu1) raring; urgency=low

  * debian/control:
    - note github.com in the package description
  * bin/ssh-import-id, ssh-import-id.1:
    - rename the --revoke option to --remove
    - we're not really "revoking" keys, but rather just removing them
      from the authorized_keys file

 -- Dustin Kirkland <kirkland@ubuntu.com>  Fri, 15 Feb 2013 15:03:53 -0600

ssh-import-id (3.11-0ubuntu1) raring; urgency=low

  * bin/ssh-import-id:
    - cleanup tempfiles left around in case we die() early

 -- Dustin Kirkland <kirkland@ubuntu.com>  Fri, 08 Feb 2013 12:35:13 -0600

ssh-import-id (3.10-0ubuntu1) raring; urgency=low

  * bin/ssh-import-id:
    - show key type in info messages

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 07 Feb 2013 17:45:18 -0600

ssh-import-id (3.9-0ubuntu1) raring; urgency=low

  * bin/ssh-import-id-lp:
    - ensure that URL is not None before using it
    - fixes exception when URL is not defined in config file and not in
      the environment

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 07 Feb 2013 11:33:29 -0600

ssh-import-id (3.8-0ubuntu1) raring; urgency=low

  * No change release; our Python and Debian/Ubuntu version numbers got
    out of sync due to a bug in my release script

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 07 Feb 2013 11:20:40 -0600

ssh-import-id (3.7-0ubuntu1) raring; urgency=low

  * bin/ssh-import-id-lp:
    - another fix for Mac OS X

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 07 Feb 2013 10:16:59 -0600

ssh-import-id (3.6-0ubuntu1) raring; urgency=low

  * setup.py:
    - fix pip install on Mac

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 07 Feb 2013 09:52:36 -0600

ssh-import-id (3.5-0ubuntu1) raring; urgency=low

  * bin/ssh-import-id, ssh-import-id.1:
    - support key revocation through a -r|--revoke option
    - only revokes keys that we label as adding (starting with this release)
  * debian/control, debian/rules, setup.py:
    - fix lintian warnings
    - clean up pypi pkginfo

 -- Dustin Kirkland <kirkland@ubuntu.com>  Wed, 06 Feb 2013 16:39:54 -0600

ssh-import-id (3.4-0ubuntu1) raring; urgency=low

  * setup.py:
    - bump version
  * debian/control, debian/rules:
    - run pep8 at build for python syntax checking
    - still trying to fix 10.04 build
  * bin/ssh-import-id-lp:
    - fix pep8 warning

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 05 Feb 2013 17:37:53 -0600

ssh-import-id (3.3-0ubuntu1) raring; urgency=low

  * debian/control, debian/rules:
    - get this building on 10.04/lucid, borrowed build logic from swift

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 05 Feb 2013 12:42:21 -0600

ssh-import-id (3.2-0ubuntu1) raring; urgency=low

  * debian/control:
    - need setuptools for compilation

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 05 Feb 2013 11:53:45 -0600

ssh-import-id (3.1-0ubuntu1) raring; urgency=low

  * debian/control:
    - ensure we have python packaging build dependencies
  * debian/control, debian/rules:
    - clean up python build dependencies, fix problem where
      dh_pysupport pulls in python-support as a dependency

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 05 Feb 2013 10:05:21 -0600

ssh-import-id (3.0-0ubuntu1) raring; urgency=low

  [ Dustin Kirkland ]
  * setup.py:
    - bump major version to 3.x since this is a complete rewrite in Python
    - add support for gh:username (github) as well as lp:username (launchpad)
  * bin/ssh-auth-id => bin/ssh-import-id, bin/ssh-auth-id-gh => bin/ssh-
    import-id-gh, bin/ssh-auth-id-lp => bin/ssh-import-id-lp, README.md,
    setup.py:
    - rename Casey's ssh-auth-id command to ssh-import-id with this merge
  * bin/ssh-import-id, bin/ssh-import-id-gh, bin/ssh-import-id-lp,
    setup.py:
    - get these pep8 clean with
       pep8 --verbose --repeat --ignore W191,E501
  * bin/ssh-import-id, bin/ssh-import-id-gh, bin/ssh-import-id-lp:
    - drop the "starts with ssh-" check
    - bad assumption, since my ecdsa key starts with "ecdsa-sha2-nistp256"
    - if it's a bad key, then ssh-keygen -l -f will shake that out
    - support wget as a fallback, when python requests module is not available
    - loop over each id in argv
    - import subprocess
  * === added directory attic, debian/control, ssh-import-id =>
    attic/ssh-import-id, ssh-import-id.py => attic/ssh-import-id.py:
    - relocate the old shell utility and the monolithic python utility to
      the attic for now; these should be removed entirely from the source
      tree in due time
  * debian/install:
    - remove binaries from the install, let setup.py handle these for us
  * bin/ssh-import-id:
    - support -o|--output - for stdout
    - return all parts of the fingerprint on validation (size, hash, name, type)
    - index into the fingerprint for the hash
    - print all parts of the fingerprint in info messages
      + this is essential to the security of ssh-import-id, for key fingerprint
        collision avoidance
    - use a helper function to print to stdout just as easily a filename
    - identify keys by their tuple, [length, hash, type]
  * debian/links, README.md, ssh-import-id.1:
    - update documentation
  * bin/ssh-import-id-lp, debian/postinst, ssh_import_id:
    - support URL environment variable and system-wide URL definition in conf
      file
    - port conf file from shell to JSON syntax
    - add a postinst script to smooth migration from ssh-import-id < 3.0
    - ignore errors in postinst script (nothing more we can do, really)
  * bin/ssh-import-id-gh, bin/ssh-import-id-lp, debian/postinst,
    ssh_import_id:
    - use wget --no-verbose rather than --quiet, so that error messages from
      wget (such as SSL cert errors) get propagated
    - use json syntax for configuration file
  * bin/ssh-import-id:

attic/ssh-import-id.py
#!/usr/bin/python
#
#    ssh-import-id - authorize a user by fetching their key
#                    from a public SSH keyserver; Launchpad.net
#                    by default
#
#    Copyright (C) 2013 Dustin Kirkland
#
#    Authors: Dustin Kirkland <dustin.kirkland@gmail.com>
#
#    Original authors of the shell implementation:
#             Dustin Kirkland <dustin.kirkland@gmail.com>
#             Scott Moser <smoser@canonical.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, version 3 of the License.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import json
import os
import subprocess
import sys
import tempfile
try:
	# Python2
	from urllib import quote_plus
except (ImportError,):
	# Python3
	from urllib.parse import quote_plus


def error(msg):
	"""Print error message on stderr and exit non-zero immediately"""
	sys.stderr.write("ERROR: %s\n" % msg)
	sys.exit(1)


def warn(msg):
	"""Print warning message on stderr but do not exit"""
	sys.stderr.write("WARNING: %s\n" % msg)


def info(msg):
	"""Print info message on stderr but do not exit"""
	sys.stderr.write("INFO: %s\n" % msg)


def valid(public_key_file, output):
	"""Validate a public key in a file"""
	# Validate that ssh can calculate the fingerprint of each key
	p = subprocess.Popen(["ssh-keygen", "-l", "-f", public_key_file], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
	stdout, stderr = p.communicate()
	# Show fingerprints on stderr (not stdout)
	sys.stderr.write(stdout.decode('utf-8'))
	sys.stderr.write(stderr.decode('utf-8'))
	if p.returncode == 0:
		# Handle output
		f = open(public_key_file, "r")
		pubkey = f.read()
		f.close()
		if output == "-":
			# Print public key to stdout
			sys.stdout.write(pubkey)
			sys.stdout.write("\n")
		else:
			# Print public key to a file
			try:
				f = open(output, "a")
				f.write("\n%s\n" % pubkey)
				f.close()
			except:
				error("Could not write to [%s]" % output)
		return True
	return False


def validate(keys, results_type, output):
	"""Use the appropriate driver for key processing"""
	if results_type == "raw":
		return validate_raw(keys, output)
	elif results_type == "json":
		return validate_json(keys, output)
	else:
		return False


def validate_raw(keys, output):
	"""Validate one or more public ssh keys in a raw file"""
	f, tmp = tempfile.mkstemp()
	os.close(f)
	rc = True
	# Split the retrieved public keys into one-line-per-file
	for k in keys.split("\n"):
		# Skip blank lines
		if k.strip():
			f = open(tmp, "w")
			f.write(k)
			f.close()
			if not valid(tmp, output):
				rc = False
	os.unlink(tmp)
	return rc


def validate_json(keys, output):
	"""Validate one or more public ssh keys in a json structure"""
	f, tmp = tempfile.mkstemp()
	os.close(f)
	rc = True
	data = json.loads(keys)
	for i in data:
		if "key" in i:
			f = open(tmp, "w")
			f.write(i["key"])
			f.close()
			if not valid(tmp, output):
				rc = False
	os.unlink(tmp)
	return rc


def get_authkeypath():
	"""Validate and return the user's authorized keys file"""
	# Only support writing to this user's authorized_keys file
	# Use the $HOME environment variable, if possible
	home = os.getenv("HOME", None)
	if home is None:
		try:
			# If $HOME wasn't set, check pwent
			home = pwd.getpwuid(os.getuid())[5]
		except:
			error("Cannot get passwd entry")
	# Ensure that the .ssh directory exists, create if necessary
	if os.direxists("%s/.ssh" % home) == False:
		os.makedirs("%s/.ssh" % home, 0o700)
	# Ensure that the authorized keys file exists, create if necessary
	if os.path.exists("%s/.ssh/authorized_keys" % home) == False:
		f = os.open("%s/.ssh/authorized_keys", os.O_WRONLY | os.O_CREAT, 0o600, "w")
		os.close(f)
	return "%s/.ssh/authorized_keys" % home


def get_default_url():
	# Default to the Launchpad URL, for legacy operation
	default_url = "https://launchpad.net/~%s/+sshkeys"
	# Allow for an override using the $URL environment variable
	url = os.getenv("URL", None)
	if url == None:
		# Source the configuration file if it exists
		if os.path.exists("/etc/ssh/ssh_import_id"):
			try:
				# UGLY: source the legacy configuration file, which was a shell variable definition
				gconf = {}
				lconf = {}
				with open("/etc/ssh/ssh_import_id") as f:
					code = compile(f.read(), "/etc/ssh/ssh_import_id", 'exec')
					exec(code, gconf, lconf)
				if "URL" in gconf:
					url = gconf["URL"]
				else:
					url = default_url
			except:
				error("Could not source /etc/ssh/ssh_import_id")
		else:
			url = default_url
	else:
		url = default_url
	return url


def get_args():
	"""Handle argument parsing"""
	parser = argparse.ArgumentParser()
	parser.add_argument("-o", "--output", help="write output to FILE; default ~/.ssh/authorized_keys, use '-' for STDOUT", default=None, metavar="FILE")
	parser.add_argument("user_id", help="USER_ID [USER_ID_2] ... [USER_ID_n]", nargs="+")
	return parser.parse_args()


def get_output(args):
	"""Determine output destination"""
	if args.output is None:
		output = get_authkeypath()
	else:
		output = args.output
	return output


def wget(url):
	"""Fetch data by url"""
	# Would love to use native Python here, but pycurl is a mess, urllib and urllib2 don't check SSL certs, and requests is available everywhere yet
	# So we're going to shell out and use good 'ole wget
	p = subprocess.Popen(["wget", "--quiet", "-O", "-", url], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
	stdout, stderr = p.communicate()
	sys.stderr.write(stderr.decode('utf-8'))
	if p.returncode == 0:
		return stdout.decode('utf-8')
	else:
		error("Could not fetch URL [%s]" % url)


def main():
	"""Main function of ssh-import-id"""
	# Define main variables
	rc = 0
	args = get_args()
	output = get_output(args)
	default_url = get_default_url()
	# Loop over user id positional arguments
	for i in args.user_id:
		# Ensure the user id is url safe (quoted)
		if i.startswith("gh:"):
			u = "https://api.github.com/users/%s/keys" % quote_plus(i.split("gh:")[1])
			results_type = "json"
		elif i.startswith("lp:"):
			u = "https://launchpad.net/~%s/+sshkeys" % quote_plus(i.split("lp:")[1])
			results_type = "raw"
		else:
			u = default_url % quote_plus(i)
			results_type = "raw"
		try:
			# Attempt to fetch the url
			results = wget(u)
		except:
			error("Unable to retrieve url [%s]" % u)
		try:
			# Validate that each non-blank line in the tempfile are good ssh keys
			if validate(results, results_type, output) == False:
				warn("Invalid keys at [%s]" % u)
				rc += 1
				continue
		except:
			error("Unable to validate keys from [%s]" % u)
		# This user's key(s) look valid!
		info("Successfully authorized [%s]" % i)
	sys.exit(rc)


if __name__ == '__main__':
	main()
